File: 1001~1500/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/main.py
"""
Source: https://rb.gy/gs3gy
Date: 2023/8/1
Skill: Floyd–Warshall algorithm
Ref:
Runtime: 582 ms, faster than 66.34%
Memory Usage: 17.36 MB, less than 97.26%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        dp, res, cnt = [[sys.maxsize // 3 for _ in range(n)] for _ in range(n)], 0, n
        for edge in edges:
            dp[edge[0]][edge[1]] = dp[edge[1]][edge[0]] = edge[2]

        for i in range(n):
            for j in range(n):
                for k in range(n):
                    dp[j][k] = min(dp[j][k], dp[j][i] + dp[i][k])

        for i in range(n):
            num = 0
            for j in range(n):
                if j != i and dp[i][j] <= distanceThreshold:
                    num += 1
            if num <= cnt:
                cnt = num
                res = i

        return res


if __name__ == "__main__":
    s = Solution()

File: 1001~1500/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/main.cpp
/**
 * Source: ibit.ly/ZSgxj
 * Date: 2023/9/28
 * Skill:
 * Runtime: 20 ms, faster than 99.10% of C++ online submissions
 * Memory Usage: 12.33 MB, less than 75.30% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *      2 <= n <= 100
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        int dis[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                dis[i][j] = INT32_MAX / 2;
        for (auto e: edges)
            dis[e[0]][e[1]] = dis[e[1]][e[0]] = e[2];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                for (int k = 0; k < n; k++) {
                    if (k != j)
                        dis[j][k] = dis[k][j] = min(dis[j][k], dis[j][i] + dis[i][k]);
                }
        int minimum = INT32_MAX, res = 0;
        for (int i = 0; i < n; i++) {
            int num = 0;
            for (int j = 0; j < n; j++) {
                if (j != i && dis[i][j] <= distanceThreshold)
                    num++;
            }
            if (num <= minimum) {
                res = i;
                minimum = num;
            }
        }
        return res;
    }
};


int main() {
    Solution s;
    vector<vector<int>> arr1{{10, 9},
                             {7,6}};
}